js/src/DappRequests/store.js

```diff
@@ -41,131 +38,137 @@ export default class Store {
-      [id]: true
-    });
-    this.savePermissions();
-  }
+    // Add a new request in this.requests
+    this.requests = {
+      ...this.requests,
+      [`${method}:${appId}`]: request
+    };
+  };
 
-  @action approveSingleRequest = ({ queueId, request: { data } }) => {
-    const source = this.sources[queueId];
+  @action approveRequest = (method, appId) => {
+    const { source, data } = this.requests[`${method}:${appId}`];
 
-    this.removeRequest(queueId);
+    this.addAppPermission(method, appId);
+    this.removeRequest(method, appId);
 
     if (data.api) {
       this.executePubsubCall(data, source);
     } else {
       this.executeMethodCall(data, source);
     }
-  }
+  };
 
-  @action approveRequest = (queueId, approveAll) => {
-    const queued = this.findRequest(queueId);
+  @action approveRequestGroup = (groupId, appId) => {
+    // Get methods of this requestGroup
+    const { methods } = methodGroups[groupId];
 
-    if (approveAll) {
-      const { request: { data: { method, token, params } } } = queued;
+    methods.forEach(method => {
+      if (!this.requests[`${method}:${appId}`]) {
+        return;
+      }
 
-      this.getFilteredSection(method || params[0]).methods.forEach((m) => {
-        this.addTokenPermission(m, token);
-        this.findMatchingRequests(m, token).forEach(this.approveSingleRequest);
-      });
-    } else {
-      this.approveSingleRequest(queued);
-    }
-  }
+      this.approveRequest(method, appId);
+    });
+  };
 
-  @action rejectRequest = (queueId) => {
-    const { request: { data } } = this.findRequest(queueId);
-    const source = this.sources[queueId];
+  @action rejectRequest = (method, appId) => {
+    const { source, data } = this.requests[`${method}:${appId}`];
 
-    this.removeRequest(queueId);
+    this.removeRequest(method, appId);
     this.rejectMessage(source, data);
-  }
+  };
+
+  @action rejectRequestGroup = (groupId, appId) => {
+    // Get methods of this requestGroup
+    const { methods } = methodGroups[groupId];
+
+    methods.forEach(method => {
+      if (!this.requests[`${method}:${appId}`]) {
+        return;
+      }
+
+      this.rejectRequest(method, appId);
+    });
+  };
+
+  @action removeRequest = (method, appId) => {
+    delete this.requests[`${method}:${appId}`];
+    this.requests = { ...this.requests };
+  };
 
   @action rejectMessage = (source, { id, from, method, token }) => {
     if (!source) {
       return;
     }
 
-    source.postMessage({
-      error: `Method ${method} not allowed`,
-      id,
-      from: 'shell',
-      result: null,
-      to: from,
-      token
-    }, '*');
-  }
+    source.postMessage(
+      {
+        error: `Method ${method} not allowed`,
+        id,
+        from: 'shell',
+        result: null,
+        to: from,
+        token
+      },
+      '*'
+    );
+  };
+
+  @action addAppPermission = (method, appId) => {
+    this.permissions = Object.assign({}, this.permissions, {
```

Comment:
```js
this.permissions = {
  ...this.permisions,
  [..]: true,
};
```
?

Is it required to copy over the permissions anyway? It might be called in a loop currently (approveGroup -> approveRequest -> addAppPermission) and it seems like a lot of overhead.

Reply:
I don't quite understand this. I'm copying this.permissions for the same reason as this.requests above, for mobx to update the store.

Reply:
Yup, sorry. I wrote the comments in opposite order :) Still replacing `Object.assign` with more idiomatic `...` pattern would look nice.

Reply:
OK, that's fixed in the last commit